src/library/list.lisp

```diff
@@ -15,13 +15,26 @@
       ((Cons x _) (Some x))
       ((Nil) None)))
 
-  (declare tail ((List :a) -> (Optional (List :a))))
+  (declare tail ((List :a) -> (List :a)))
   (define (tail l)
-    "Returns every element but the first in a list."
+    "Returns every element except the first in a list."
     (match l
-      ((Cons _ xs) (Some xs))
+      ((Cons _ xs) xs)
+      ((Nil) Nil)))
+
+  (declare last ((List :a) -> (Optional :a)))
+  (define (last l)
+    "Returns the last element of a list."
+    (match l
+      ((Cons x (Nil)) (Some x))
+      ((Cons _ xs) (last xs))
       ((Nil) None)))
 
+  (declare init ((List :a) -> (List :a)))
+  (define (init l)
+    "Returns every element except the last in a list."
+    (reverse (tail (reverse l))))
```

Comment:
CL:BUTLAST could also be used

Reply:
Should I use CL:BUTLAST here?  I have no preference.

Reply:
It will be about 2x as efficient, but it doesn't matter. (It'll just make one copy, not two.